### PR TITLE
Test ma57_alter_d

### PR DESCRIPTION
**Pull request by amontoison**
`ma57_alter_d` seems to not working :(

### CODE REVIEW

File: test/test_hsl_ma57_patch.jl [hunk 1]
```diff
@@ -26,6 +26,8 @@ function test_ma57_patch(A, M, b, xexact)
   d1 = abs.(diag(D))
   d2 = [diag(D, 1); 0][:]
   ma57_alter_d(M, [Vector(d1)'; Vector(d2)'])
+  x4 = ma57_solve(M, b)
+  @test norm(x4 - xexact) ≤ ϵ * norm(xexact)
```
Comment (dpo) [inline test/test_hsl_ma57_patch.jl:30]: This couldn't possibly hold since you've modified the factors.

File: test/test_hsl_ma57_patch.jl [hunk 2]
```diff
@@ -26,6 +26,8 @@ function test_ma57_patch(A, M, b, xexact)
   d1 = abs.(diag(D))
```
Comment (amontoison) [inline test/test_hsl_ma57_patch.jl:26]:
```suggestion
  d1 = diag(D)
```

File: test/test_hsl_ma57_patch.jl (hunk 1 above)
Comment (amontoison) [reply test/test_hsl_ma57_patch.jl:30]: Oups, I updated the test.
I checked the Julia code and it's a copy of the Fortran code.
Tjey probably never tested this routine or only with matrices with 1x1 or 2x2 blocks.

Comment (dpo) [reply test/test_hsl_ma57_patch.jl:30]: If you change the factors, you no longer have a factorization of A. So there's no reason to expect that x4 would be close to xexact. It's the solution to a different system.

Comment (dpo) [reply test/test_hsl_ma57_patch.jl:30]: Ah sorry, I did not see the updated test. Yes, that's a problem.

Comment (dpo) [reply test/test_hsl_ma57_patch.jl:30]: If you extract D again, what's inside?

Comment (amontoison) [reply test/test_hsl_ma57_patch.jl:30]: I found the problem, `ma57_inter_d` expect that you provide `D^{-1}` and not `D`...

```julia
using HSL

T = Float64
rows = Cint[1, 1, 2, 2, 3, 3, 5]
cols = Cint[1, 2, 3, 5, 3, 4, 5]
vals = T[2, 3, 4, 6, 1, 5, 1]
A = sparse(rows, cols, vals)
A = A + triu(A, 1)'

M = ma57_coord(5, rows, cols, vals)

b = T[8, 45, 31, 15, 17]
xexact = T[1, 2, 3, 4, 5]

ϵ = sqrt(eps(eltype(A)))
ma57_factorize!(M)

# extract factors
(L, D, s, p) = ma57_get_factors(M)

invD = inv(Matrix(D))
d1 = diag(invD)
d2 = [diag(invD, 1); 0][:]
Dnew = [Vector(d1)'; Vector(d2)']

ma57_alter_d(M, Dnew)

(L2, D2, s2, p2) = ma57_get_factors(M)

@test norm(D - D2) ≤ ϵ

x4 = ma57_solve(M, b)
@test norm(x4 - xexact) ≤ ϵ * norm(xexact)
```

**Comment by amontoison** [comment]
@AntoninKns 
Can you test that `minres_qlp` converges in a few iterations if you use MA57 as preconditioner with our modified block diagonal matrix?
You just need to add `M = inv(Tridiagonal(M)` in your code before `d1 = diag(M)`.

File: test/test_hsl_ma57_patch.jl (hunk 1 above)
Comment (amontoison) [reply test/test_hsl_ma57_patch.jl:31]: I update the docstrings.
Dominique, they added your patch in the release 5.3.2 for information.

Comment (dpo) [reply test/test_hsl_ma57_patch.jl:31]: Phew! Yes, I thought my patch made it into the latest release. As a backup, I mentioned to @AntoninKns that the functionality also comes standard in MA97 (though I forgot to document it).

**Review by dpo — APPROVED** [review]

**Comment by AntoninKns** [comment]
I tried testing it with the following code:
```julia
LDL = Ma57(A)
ma57_factorize!(LDL)
(L, D, s, p) = ma57_get_factors(LDL)
invD = inv(Tridiagonal(D))
d1 = diag(invD)
d2 = [diag(invD, 1) ; 0][:]
F = [Vector(d1)' ; Vector(d2)']
ma57_alter_d(LDL, F)
```
But it seems I don't have enough RAM to calculate the inverse:
```julia
ERROR: LoadError: OutOfMemoryError()
Stacktrace:
  [1] Array
    @ ./boot.jl:459 [inlined]
  [2] Array
    @ ./boot.jl:467 [inlined]
  [3] zeros
    @ ./array.jl:525 [inlined]
  [4] Matrix{Float64}(s::LinearAlgebra.UniformScaling{Bool}, dims::Tuple{Int64, Int64})
    @ LinearAlgebra /apps/local-fci/tools/julia-1.7.1/share/julia/stdlib/v1.7/LinearAlgebra/src/uniformscaling.jl:484
  [5] Array
    @ /apps/local-fci/tools/julia-1.7.1/share/julia/stdlib/v1.7/LinearAlgebra/src/uniformscaling.jl:491 [inlined]
  [6] inv(A::LinearAlgebra.Tridiagonal{Float64, SparseArrays.SparseVector{Float64, Int32}})
    @ LinearAlgebra /apps/local-fci/tools/julia-1.7.1/share/julia/stdlib/v1.7/LinearAlgebra/src/generic.jl:1070
  [7] solve_sub_problem!(model::BundleAdjustmentModel{Float64, Vector{Float64}}, generic_solver::LDLSolver{Float64, Vector{Float64}, Float64}, in_axtol::Float64, in_btol::Float64, in_atol::Float64, in_rtol::Float64, in_etol::Float64, in_itmax::Int64, in_conlim::Float64, #unused#::Val{false})
    @ LevenbergMarquardt ~/LevenbergMarquardt.jl/src/LMProblem.jl:275
```
And I tested it on one of the smallest problems :(

**Review by AntoninKns — APPROVED** [review]

**Comment by AntoninKns** [comment]
Since the tests are working the correction you made is still right but I will try to make another function that doesn't require calculating the inverse of D everytime.

**Comment by amontoison** [comment]
I tested and `inv(Tridiagonal(M))` is not optimized, it returns a dense matrix...
I understand you `OutOfMemoryError`.
You need to implement a function that returns the inverse of your block diagonal but it's easy.
For 1x1 block c, you just need `1/c` and for 2x2 matrix, you have the following relation:
Capture d’écran du 2022-11-12 00-06-50

**Pull request merged**